presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -278,9 +286,21 @@ private Type getCommonSuperTypeForDecimal(DecimalType firstType, DecimalType sec
         return createDecimalType(targetPrecision, targetScale);
     }
 
-    private Type getCommonSuperTypeForVarchar(VarcharType firstType, VarcharType secondType)
+    private Type getCommonSuperTypeForCharacterStrings(Type firstType, Type secondType)
+    {
+        int returnTypeLength = Math.max(getCharacterStringTypeLength(firstType), getCharacterStringTypeLength(secondType));
+        if (firstType instanceof VarcharType || secondType instanceof VarcharType) {
+            return createVarcharType(returnTypeLength);
+        }
+        return createCharType(returnTypeLength);
```

Comment:
I think `longer -> shorter` is only true when the trailing characters are spaces: "If the source is larger than the target, but the extra characters are all spaces, the source's significant character string value is assigned to the target. If the source is larger than the target and the extra characters are not all spaces, the assignment will fail"